### PR TITLE
Add Windows CI

### LINKED ISSUE #11434: [Bug] Unexplored Windows CI failures
### What happened?

There are failures in the Windows CI PR that have not been attributed to things we don't care about.  Might be worth making sure we know what they are and whether they are relevant to 1.3.5.

https://github.com/Chia-Network/chia-blockchain/pull/10681

### Version

main

### What platform are you using?

Windows

### What ui mode are you using?

CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by altendky**
Draft for:
- [x] Self review and tidying
- [x] Presumably various issues to address
- [x] Check the Python version used in envs, seems to be 3.10, probably due to versionless `py` call in `Install.ps1`.
- [x] Revisit `--numprocesses 0` in case xdist is usable on Windows.
- [x] Windows server test failures.
  - https://github.com/aio-libs/aiohttp/issues/2934
- [x] Break out PRs:
  - https://github.com/Chia-Network/chia-blockchain/pull/10940
  - https://github.com/Chia-Network/chia-blockchain/pull/11117
- [ ] Maybe get timelord going...

Items to note.

The build workflow script currently sets `-n 0` for all windows tests and ignores the config `Parallel` setting. Windows tests do work locally quite well with multiple subprocesses, however, the GH runners stumble with 4 processes in some tests. Subsequent PRs can explore expanding the config to have a dict with per OS settings.

The IPv6 test is skipped (also skipped on macOS).

### CODE REVIEW

File: .github/workflows/build-test-windows-blockchain.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    - name: Checkout test blocks and plots
+      uses: actions/checkout@v2
+      with:
+        repository: 'Chia-Network/test-cache'
+        path: '.chia'
+        ref: '0.28.0'
+        fetch-depth: 1
+
+    - name: Link home directory
```
Comment (arvidn) [inline .github/workflows/build-test-windows-blockchain.yml:73]: this step has unixy stuff in it. Is it necessary to do on windows?

File: .github/workflows/build-test-windows-blockchain.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    - name: Run install script
+      env:
+        INSTALL_PYTHON_VERSION: ${{ matrix.python-version }}
+      run: |
+        ./Install.ps1
+
+    - name: Install timelord
+      run: |
+        . ./activate
+        sh install-timelord.sh
```
Comment (arvidn) [inline .github/workflows/build-test-windows-blockchain.yml:89]: I would expect this to be problematic on windows too.

File: .github/workflows/build-test-windows-blockchain.yml (hunk 1 above)
Comment (altendky) [reply .github/workflows/build-test-windows-blockchain.yml:73]: I started to poke at alternatives over in https://github.com/Chia-Network/chia-blockchain/pull/10682.

File: tests/pools/test_pool_config.py [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
 from chia.pools.pool_config import PoolWalletConfig
-from chia.util.config import load_config, lock_config, save_config, create_default_chia_config
+from chia.util.config import create_default_chia_config, load_config, lock_config, save_config
 
 
 def test_pool_config():
-    test_root = Path("/tmp")
-    test_path = Path("/tmp/config")
+    test_root = Path(gettempdir())
+    test_path = test_root / "config"
```
Comment (altendky) [inline tests/pools/test_pool_config.py:13]: This is a step forward, but tests should be getting a temporary directory from the `tmp_path` fixture.  This will provide a per-test temporary directory where all of those directories in a given run will be inside the pytest temporary directory for that run which are all inside the overall pytest temporary directory.  We definitely should never have put a hard coded file name directly in `/tmp`, nor even the system temp directory.

File: tests/core/server/test_dos.py [hunk 4]
```diff
@@ -201,9 +201,15 @@ async def test_spam_tx(self, setup_two_nodes_fixture, self_hostname):
...
-        await asyncio.sleep(1)
+        try:
+            for i in range(6000):
+                await ws_con._send_message(new_tx_message)
+            await asyncio.sleep(1)
+        except ConnectionResetError:
+            # based on timing the connection may get closed by rate limits during the loop
+            # and this is actually want we want so catch this exception and ignore it
+            # seen primarily on my Windows box
+            pass
```
Comment (altendky) [inline tests/core/server/test_dos.py:212]: Typo in `want we want`.  The sleep could go outside as well, I think?

Why do we not get the exception everywhere else?  This seems like a good thing to understand a bit better.

**Review by altendky — COMMENTED** [review]
Thanks for the work and for finding some of the issues here.

File: tests/core/server/test_dos.py (hunk 4 above)
Comment (emlowe) [reply tests/core/server/test_dos.py:212]: The rate limits are x messages within x timeframe - we actually want it to fire sometime here in the loop. I kinda would have expected this exception in more places actually.

File: tests/core/test_farmer_harvester_rpc.py [hunk 5]
```diff
@@ -435,7 +435,7 @@ def test_plot_matches_filter(filter_item: FilterItem, match: bool):
         ),
         (FarmerRpcClient.get_harvester_plots_invalid, [], None, True, 13),
         (FarmerRpcClient.get_harvester_plots_invalid, ["invalid_0"], None, False, 6),
-        (FarmerRpcClient.get_harvester_plots_invalid, ["inval", "lid_1/"], None, False, 2),
+        (FarmerRpcClient.get_harvester_plots_invalid, ["inval", "lid_1"], None, False, 2),
```
Comment (altendky) [inline tests/core/test_farmer_harvester_rpc.py:438]: @xdustinface, is the need to remove the `/` here a sign of some lack of normalization of the paths or something?  @emlowe, could you comment on how this was failing?  Was it `/` vs. `\`?  or just no trailing slash to match at all?

Comment (emlowe) [reply tests/core/test_farmer_harvester_rpc.py:438]: Yea, it's a string \ vs / - I dunno if there is a better way to build the test, but removing the / was what Kevin suggested

Comment (altendky) [reply tests/core/test_farmer_harvester_rpc.py:438]: I'm definitely tempted to normalize them for comparison, though I haven't thought through corner cases.  `/` is generally valid all over Windows.  But, that would be a feature addition so not really required here.  But, might be worth xfailing this on Windows?  This suggestion needs at least an `import sys` added so apply this manually.

pytest.param(1, 3, marks=pytest.mark.xfail(reason="some bug")),
```suggestion
        pytest.param(
            FarmerRpcClient.get_harvester_plots_invalid,
            ["inval", "lid_1/"],
            None,
            False,
            2,
            marks=pytest.mark.xfail(condition=sys.platform == "win32", reason="not presently supported on Windows"),
        ),
        pytest.param(
            FarmerRpcClient.get_harvester_plots_invalid,
            ["inval", "lid_1\\"],
            None,
            False,
            2,
            marks=pytest.mark.skip(condition=sys.platform != "win32", reason="only for Windows"),
        ),
```

Comment (xdustinface) [reply tests/core/test_farmer_harvester_rpc.py:438]: @altendky It's good enough to just drop the `/`. It's just filtering for the path there, not really important to test that path filtering for windows is with `\` and for others with `/` imo. I initially just added it to avoid flakiness with the temp directory names because i thought it might contain `lid_1` but well let's see how that goes?

**Review by altendky — COMMENTED** [review]
I don't like all the `time_out_assert()` time increases, but maybe we just accept that and focus instead on https://github.com/Chia-Network/chia-blockchain/pull/11535 and then applying it more.

File: tests/wallet/test_wallet_retry.py [hunk 6]
```diff
@@ -42,6 +44,10 @@ def evict_from_pool(node: FullNodeAPI, sb: SpendBundle) -> None:
 
 
 @pytest.mark.asyncio
+@pytest.mark.skipif(
+    condition=("GITHUB_ACTIONS" in os.environ) and (sys.platform in {"doodad"}),
+    reason="testing stuff",
+)
```
[the remark this answers is not in the document]
Comment (emlowe) [reply tests/wallet/test_wallet_retry.py:50]: oh, ya, I forgot to remove that !

File: tests/core/server/test_dos.py (hunk 4 above)
Comment (mariano54) [reply tests/core/server/test_dos.py:212]: The connection does close, that is checked in line 217. So it must be one of two scenarios:
1. On other platforms, there is a delay when closing connection
2. On other platforms, you can still call ws_con.send_message on a closed WS.

I'm not sure which is the right one

Comment (altendky) [reply tests/core/server/test_dos.py:212]: https://github.com/Chia-Network/chia-blockchain/pull/11612

Comment (emlowe) [reply tests/core/server/test_dos.py:212]: I was thinking of the tests could be rewritten a little bit - we can use the code in rate_limits.py to use the exact number of messages that should send it over the limit - rather than this rather obscure 6000. I believe the frequency limit for NewTransaction is 5000. This is also testing the inbound limit, so the other end needs to close the connection - I suspect Windows ends up closing the connection much faster than Unix - so my guess is it's (1). That said, I'm not sure how much it matters to track it down exactly.

**Comment by mariano54** [comment]
I think we should consider using -n 4 and just commenting out the tests which don't work (or attempt to fix them).
I rather have no windows tests and 20min CI, than windows tests and 40min CI.

File: tests/core/server/test_dos.py (hunk 4 above)
Comment (mariano54) [reply tests/core/server/test_dos.py:212]: I think with kyle's fix we are good, there is no need to spend more time on this

Comment (altendky) [reply tests/core/server/test_dos.py:212]: the tests are flaky anyways.  unless mariano ended up handling this.  i had started working on that in https://github.com/Chia-Network/chia-blockchain/pull/9206 a long time ago.  but that's just general, not about windows.

File: tests/build-workflows.py [hunk 7]
```diff
@@ -146,7 +146,13 @@ def dir_path(string):
             logging.info(f"Skipping {dir}: no tests collected")
             continue
         conf = update_config(module_dict(testconfig), dir_config(dir))
+        if conf["install_timelord"] and os == "windows":
+            # TODO: enable timelord for windows
+            continue
         replacements = generate_replacements(conf, dir)
+        if os == "windows":
+            # TODO: force windows to sequential only
```
[the remark this answers is not in the document]
Comment (altendky) [reply tests/build-workflows.py:154]:
```suggestion
            # TODO: review for potential stable level of parallel testing processes
```

**Review by 1-justinstarr — COMMENTED** [review]
for merge into win ci

File: tests/core/server/test_dos.py (hunk 4 above)
Comment (altendky) [reply tests/core/server/test_dos.py:212]: Undoing changes here to allow https://github.com/Chia-Network/chia-blockchain/pull/11612 to fix it when it gets merged.

```suggestion
        for i in range(6000):
            await ws_con._send_message(new_tx_message)
        await asyncio.sleep(1)
```

**Comment by emlowe** [comment]
Ok various experiments leave me to conclude that using 2 xdist processes (`-n 2`) results in the best performance on the Windows runners. The longest test is the core-full-node tests which take 25-30 mins. Using more than 2 processes seems to increase the run time and increase flakes (due to timeouts being hit).
To further improve run times, we would likely need to break up the core-full-node tests into multiple jobs

File: tests/build-workflows.py [hunk 8]
```diff
@@ -75,9 +75,15 @@ def generate_replacements(conf, dir):
         "TEST_FILES": "",
         "TEST_NAME": "",
         "PYTEST_PARALLEL_ARGS": "",
+        "RUN": "true",
     }
 
-    xdist_numprocesses = {False: 0, True: 4}.get(conf["parallel"], conf["parallel"])
+    if os == "windows":
+        # TODO: review for potential stable level of parallel testing processes
```
Comment (altendky) [inline tests/build-workflows.py:82]:
```suggestion
    # TODO: design a configurable system for this
    if os == "windows":
```

File: tests/testconfig.py [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
+from typing import List, Union
 
-if TYPE_CHECKING:
-    from typing_extensions import Literal
+from typing_extensions import Literal
 
+Oses = Literal["macos", "ubuntu", "windows"]
 # Github actions template config.
-oses = ["ubuntu", "macos"]
+oses: List[Oses] = ["ubuntu", "macos", "windows"]
```
Comment (altendky) [inline tests/testconfig.py:9]: Really wish this was `linux`, but i'll let this comment satisfy my desires for now.

```suggestion
oses: List[Oses] = ["macos", "ubuntu", "windows"]
```

**Review by emlowe — DISMISSED** [review]
I approve

File: tests/pools/test_pool_rpc.py [hunk 10]
```diff
@@ -269,8 +269,8 @@ async def test_create_new_pool_wallet_farm_to_pool(self, one_wallet_node_and_rpc
 
         await farm_blocks(full_node_api, our_ph, 6)
         assert full_node_api.full_node.mempool_manager.get_spendbundle(creation_tx.name) is None
-
         await time_out_assert(20, wallet_is_synced, True, wallet_node_0, full_node_api)
+
```
Comment (xdustinface) [inline tests/pools/test_pool_rpc.py:273]: nit: unrelated?

File: tests/runner_templates/checkout-test-plots.include.yml [hunk 11]
```diff
@@ -7,9 +7,16 @@
           ${{ github.workspace }}/.chia/test-plots
         key: 0.29.0
 
-    - name: Checkout test blocks and plots
-      if: steps.test-blocks-plots.outputs.cache-hit != 'true'
+    - name: Checkout test blocks and plots (Linux, macOS)
+      if: steps.test-blocks-plots.outputs.cache-hit != 'true' && (startswith(matrix.os, 'ubuntu') || startswith(matrix.os, 'macos'))
```
Comment (xdustinface) [inline tests/runner_templates/checkout-test-plots.include.yml:11]: Wouldn't it be possible to merge those into one step and have different run sections depending on the OS? Doesn't matter much just thought it might be cleaner in the action overview page :)

**Review by xdustinface — DISMISSED** [review]
Looks good, two nits to maybe address, or not :)

File: tests/pools/test_pool_rpc.py (hunk 10 above)
Comment (altendky) [reply tests/pools/test_pool_rpc.py:273]: see 9fc881f0ae5787c030be1b92421a4c503a5ed114

File: tests/runner_templates/checkout-test-plots.include.yml (hunk 11 above)
Comment (xdustinface) [reply tests/runner_templates/checkout-test-plots.include.yml:11]: Got it. I just thought there might be some way to do something like 

```
- name: Checkout test blocks and plots
   if: macos/linux
   run: ....
   if windows:
   run: ....
```

But i don't know much about github actions :)

**Review by xdustinface — CHANGES_REQUESTED** [review]
The LGTM alert indeed was introduced here because you removed the usage of `Path` in `tests/pools/test_pool_config.py` and it seems like i made the Windows CI unhappy again with some path related issues via #11625, sorry :) Seems like its related to the path string matching in `with pytest.raises(ValueError, match=f"Path doesn't exist: {test_path}")`?

**Review by xdustinface — APPROVED** [review]

**Review by wjblanke — APPROVED** [review]
aok
